### ISSUE TITLE
dump and restore functionality for in memory storage

### ISSUE THREAD

**Issue by sido420** [animir/node-rate-limiter-flexible#40]
I was thinking if we could add a dump/restore functionality to this module. 

Due to the nature of nodejs process, if it goes down we loose everything stored in the in-memory db.

What are your thoughts?

**Comment by animir** [comment]
@sido420 This is interesting idea. It would be definitely useful to have such feature in some cases.

**Comment by sido420** [comment]
It came from the popular LRU cache nodejs module, see https://github.com/isaacs/node-lru-cache. 

I think it'd be a huge benefit for single node setups where external storage is unavailable. Or, even multi process setups with sticky sessions etc. or even without sticky sessions it might be of help. Quick app prototypes should also benefit from it :)

**Comment by animir** [comment]
@sido420 Yeah, I think so. Community would appreciate it. Let us know if you started to implement it, so nobody else does the same in parallel. Thanks.

**Comment by sido420** [comment]
No, I haven't started any work on it. 

I will comment here if I plan on doing so (not anytime soon).

**Comment by sido420** [comment]
I stringified _storage in memory storage and got the following:
```
{"_storage":{"Msg-ByID:1":{"_value":1,"_expiresAt":"2019-10-08T00:59:58.025Z","_timeoutId":{"_called":false,"_idleTimeout":600000,"_idlePrev":null,"_idleNext":null,"_idleStart":229
:/conta│15556,"_repeat":null,"_destroyed":false,"_handle":{}}}}}
```
Can you tell me what needs to persist so it could be restored at startup?

**Comment by animir** [comment]
@sido420 `_timeoutId` contains data we can't use really. Every key should re-create a timer with `setTimeout` function depending on `_expiresAt` and set `_timeoutId`. So you could remove `_timeoutId` from dump.

**Comment by man610** [comment]
Is some one working on this?

**Comment by animir** [comment]
@man610 Hi, nobody works on this feature. Would you like to implement it?

**Comment by man610** [comment]
I came across this and found it interesting. I’d be happy to give it a try.

**Comment by animir** [comment]
@man610 Ok, I assigned it to you. Do not hesitate to ask questions if something is not clear about memory limiter.

**Comment by man610** [comment]
I just confused around something do we depends on other package like lru-cache we can have something like file logs which we store to system or somewhere ?
can you guide me through it or can you share some doc which i can go through it for help
sorry it can be look something annoying

happy to get positive response and guide i am super excited for my first open source contribution

**Comment by animir** [comment]
@man610 You should implement a couple of methods on RateLimiterMemory to dump data and restore data in JSON format. Take a look at how RateLimiterMemory works.

Here is short TODO list:
1. **Add `dumpToString()` to `RateLimiterMemory`**
   Serializes `_storage` + `_inMemoryBlockedKeys` with absolute `expiresAt` (ms epoch) and `version: 1`. Returns a JSON as string.

2. **Add `restoreFromString(serialized)` to `RateLimiterMemory`**
   Parses JSON string, checks `version === 1`, drops expired records, reconstructs `setTimeout` with `expiresAt - Date.now()` for survivors. Fails on JSON parse error or dump version mismatch.

3. **Add internal `_restoreRecord(key, value, expiresAt)` on `MemoryStorage`** Marked `@internal`.

4. **Include `_inMemoryBlockedKeys` in the serialization**
   Otherwise blocked keys get unblocked on restart.

5. **TypeScript definitions**
   `dumpToString(): string` and `restoreFromString(serialized: string): void`. No new exports.

6. **Write Unit Tests**
   Dump, restore, full-cycle, TTL arithmetic across simulated downtime (shorter and longer than TTL), blocked keys survival, corrupt input → no throw, version mismatch → no throw.

Export format example:
```json
{
  "version": 1,
  "dumpedAt": 1744800000000,
  "storage": [
    { "key": "rlflx:192.168.1.1", "value": 9, "expiresAt": 1744800030000 }
  ],
  "blockedKeys": [
    { "key": "rlflx:192.168.1.2", "expiresAt": 1744800900000 }
  ]
}
```

Let me know if that helps.

**Comment by man610** [comment]
will it be look something like this or we need to write file in our own package 

My previous thought is without extra work on client side to setup we just store that internally  and restore silently what is your thought about this? 

const fs = require('fs');
const { RateLimiterMemory } = require('rate-limiter-flexible');

const rateLimiter = new RateLimiterMemory({
  points: 10,
  duration: 1,
});

// --- On Startup ---
if (fs.existsSync('./ratelimiter-dump.json')) {
  const data = fs.readFileSync('./ratelimiter-dump.json', 'utf8');
  rateLimiter.restoreFromString(data);
}

// --- On Shutdown ---
process.on('SIGTERM', () => {
  const dump = rateLimiter.dumpToString();
  fs.writeFileSync('./ratelimiter-dump.json', dump);
  process.exit(0);
});

**Comment by animir** [comment]
@man610 RateLimiterMemory provides methods to dump and restore data. We shouldn't implement any logic on where to store that data. Some users may want to store it to file, others will store it to database or browser's local storage.

We can put your example of `On Startup` and `On Shutdown` logic to docs.

**Comment by man610** [comment]
check pr 
https://github.com/animir/node-rate-limiter-flexible/pull/361

and suggest changes if there is any i am not that damm sure

**Comment by animir** [comment]
@man610 I updated [RateLimiterMemory doc](https://github.com/animir/node-rate-limiter-flexible/wiki/Memory).
Please, take a look.

**Comment by animir** [comment]
Released on v11.1.0
